### PR TITLE
MVP One

### PR DESCRIPTION
**Pull request by vaskalan**
We need to discuss some issues after review.

### CODE REVIEW

File: dataset_collector/app.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        # Let's think it over - never ending loop may occur when job service fails
+        # and does not clean input directory.
+        #
+        # Example:
+        # Error with "message": "Job with target VARIABLE_3 already in progress" keeps logging without removing
+        # VARIABLE_3 from the pipeline's input dir and this triggers VARIABLE_3 to be repported by
+        # job service /importable-datasets endpoint again and again.
+
+        except Exception as exc:
```
Comment (DanielElisenberg) [inline dataset_collector/app.py:41]: catching all exceptions inside of the while True will ensure that it keeps running no matter what happens 😱 This should catch named exceptions where we expect the loop to continue, and other exception handling should be done outside of the loop 👍🏻

File: dataset_collector/app.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+json_logging.init_non_web(custom_formatter=CustomJSONLog, enable_json=True)
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+logger.addHandler(logging.StreamHandler(sys.stdout))
+
+RUN_JOB_SERVICE_CLIENT = environment.get('RUN_JOB_SERVICE_CLIENT')
+
+
+def main():
+    log_thread = threading.Thread()
```
Comment (DanielElisenberg) [inline dataset_collector/app.py:22]: As we are not starting any child processes we do not need the logger thread from the executor loop👍🏻

File: SECURITY.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Security Policy
```
Comment (DanielElisenberg) [inline SECURITY.md:1]: :100:

File: README.md [hunk 4]
```diff
@@ -1,2 +1,29 @@
 # microdata-dataset-collector
-Collects and decrypts incoming datasets
+
+#### Collects and decrypts incoming datasets.
+Detects encrypted files in ```ENCRYPTED_DIR``` and decrypts them into ```DECRYPTED_DIR```
+
+Two extensions of files are expected:
```
Comment (pawbu) [inline README.md:6]: No mention about the .json file?

File: dataset_collector/adapter/job_service_client.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    if response.status_code != 200:
+        raise HttpResponseError(f'{response.text}')
+
+    logger.info(f'Post request for jobs submitted. Response: {json.dumps(response.json(), indent=2)}')
+
+
+def import_datasets():
+    jobs = __create_jobs()
+    if len(jobs) > 0:
+        __import_request(__create_jobs())
```
Comment (pawbu) [inline dataset_collector/adapter/job_service_client.py:52]: `_import_request(jobs)`

File: dataset_collector/adapter/job_service_client.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+logger = logging.getLogger()
+
+JOB_SERVICE_URL = os.environ['JOB_SERVICE_URL']
+
+IMPORTABLE_DATASETS_URL = f'{JOB_SERVICE_URL}/importable-datasets'
+JOBS_URL = f'{JOB_SERVICE_URL}/jobs'
+
+
+def __create_jobs() -> List:
```
Comment (pawbu) [inline dataset_collector/adapter/job_service_client.py:19]: Use single underscore for "private" methods.

File: dataset_collector/app.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    while True:
+        try:
+            time.sleep(10)
+
+            decrypt_service.decrypt()
+        #   if RUN_JOB_SERVICE_CLIENT == "true":
+        #      job_service_client.import_datasets()
+
+        # Let's think it over - never ending loop may occur when job service fails
+        # and does not clean input directory.
```
Comment (pawbu) [inline dataset_collector/app.py:34]: Yes, the datasets that fail will have to be moved to a separate directory for manual checking.

File: dataset_collector/decrypt/decrypt_service.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+logger = logging.getLogger()
+
+RSA_KEY_DIR = environment.get('RSA_KEY_DIR')
+ENCRYPTED_DIR = environment.get('ENCRYPTED_DIR')
+DECRYPTED_DIR = os.environ['DECRYPTED_DIR']
+
+
+def decrypt():
+    rsa_key_dir = Path(RSA_KEY_DIR)
+    if not rsa_key_dir.exists():
```
Comment (DanielElisenberg) [inline dataset_collector/decrypt/decrypt_service.py:24]: These checks are good :100: They can be moved to the module root level so the app fails on startup if these are not set 👍🏻

File: dataset_collector/util/file_service.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import logging
```
Comment (DanielElisenberg) [inline dataset_collector/util/file_service.py:1]: We have usually put this module as 'adapter/local_storage.py'

File: tests/unit/decrypt/test_decrypt_service.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    shutil.rmtree(ENCRYPTED_DIR)
+    shutil.rmtree(ARCHIVE_DIR)
+
+
+def test_decrypt():
+    decrypt_service.decrypt()
+
+    actual = f'{DECRYPTED_DIR}/VARIABLE_1/VARIABLE_1.csv'
+    expected = Path('tests/resources/VARIABLE_1_expected.csv')
+    assert filecmp.cmp(actual, expected)
```
Comment (DanielElisenberg) [inline tests/unit/decrypt/test_decrypt_service.py:27]: :100:

**Review by DanielElisenberg — COMMENTED** [review]
I am wondering how quick the transfer from the filezilla is, and if we are risking starting a decryption mid-transfer on a partially done file 🤔 ?

Maybe we can fetch the size in the loop of a file we have not seen before like:
```py
os. path. getsize('file_path')
```
and then on the next loop check if the size is the same? Then we can know with some certainty that the filetransfer is done, and it is safe to start decrypting.

File: dataset_collector/app.py (hunk 7 above)
Comment (vaskalan) [reply dataset_collector/app.py:34]: I disconnected job_service_client from the loop as discussed.

File: dataset_collector/decrypt/decrypt_service.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+                # decrypt csv file
+                with open(f'{encrypted_dir}/{csv_file}', 'rb') as f:
+                    data = f.read()
+
+                fernet = Fernet(decrypted_symkey)
+                try:
+                    decrypted = fernet.decrypt(data)
+                    with open(f'{encrypted_dir}/{variable_name}.csv', 'wb') as f:
+                        f.write(decrypted)
+                except InvalidToken as e:
```
Comment (DanielElisenberg) [inline dataset_collector/decrypt/decrypt_service.py:78]: Might be cleaner to catch this in the outer try 🤔

File: README.md [hunk 12]
```diff
@@ -1,2 +1,32 @@
 # microdata-dataset-collector
-Collects and decrypts incoming datasets
+
+#### Collects and decrypts incoming datasets.
+Detects encrypted files in ```ENCRYPTED_DIR``` and decrypts them into ```DECRYPTED_DIR```
+wich will typically be configured as the input directory of the import pipeline.
```
Comment (DanielElisenberg) [inline README.md:5]: small typo 'wich' => 'which'

**Review by DanielElisenberg — APPROVED** [review]
MVP LGTM :100: minor comments

File: dataset_collector/decrypt/decrypt_service.py [hunk 13]
```diff
@@ -47,12 +46,13 @@ def decrypt():
 
         for csv_file in csv_files:
             try:
+                _check_for_EOF(csv_file=f'{encrypted_dir}/{csv_file}')
```
Comment (pawbu) [inline dataset_collector/decrypt/decrypt_service.py:49]: Nice with some docs explaining what is happening here.

**Review by pawbu — APPROVED** [review]
Looks good!

**Pull request merged**